modules/cnn_3dobj/include/opencv2/cnn_3dobj.hpp

```diff
@@ -0,0 +1,252 @@
+using std::string;
+using caffe::Blob;
+using caffe::Caffe;
+using caffe::Datum;
+using caffe::Net;
+/** @defgroup cnn_3dobj 3D object recognition and pose estimation API
+
+As CNN based learning algorithm shows better performance on the classification issues,
+the rich labeled data could be more useful in the training stage. 3D object classification and pose estimation
+is a jointed mission aimming at seperate different posed apart in the descriptor form.
+
+In the training stage, we prepare 2D training images generated from our module with their
+class label and pose label. We fully exploit the information lies in their labels
+by using a triplet and pair-wise jointed loss function in CNN training.
+Both class and pose label are in consideration in the triplet loss. The loss
+will be smaller when features from the same class and same pose is more similar
+and features from different classes or different poses are much larger.
+This loss is also jointed with a pair wise componant to make sure the loss is never be zero
+and have a restriction on the model scale.
+
+About the training and feature extraction process, it is a rough implemetation by using OpenCV
+and Caffe from the idea of Paul Wohlhart. The principal purpose of this API is construting
+a well labeled database from .ply models for CNN training with triplet loss and extracting features
+with the constructed model for prediction or other purpose of pattern recognition, algorithms into two main Class:
+
+**icoSphere**: methods belonging to this class generates a well generated database for CNN
+training together with their class and pose from camera view labels.
+
+**extractor**: methods belonging to this class extract features from 2D images which is
+discriminant on catogery prediction and pose estimation.
+
+@note This API need Caffe with triplet version which is designed for this module
+<https://github.com/Wangyida/caffe/tree/cnn_triplet>.
+
+*/
+namespace cv
+{
+namespace cnn_3dobj
+{
+
+//! @addtogroup cnn_3dobj
+//! @{
+
+/** @brief Icosohedron based camera view data generator.
+ The class create some sphere views of camera towards a 3D object meshed from .ply files @cite hinterstoisser2008panter .
+ */
+
+/************************************ Data Generation Class ************************************/
+    class CV_EXPORTS_W icoSphere
+    {
+        private:
+        /** @brief X position of one base point on the initial Icosohedron sphere,
+          Y is set to be 0 as default.
+         */
+        float X;
+        /** @brief Z position of one base point on the initial Icosohedron sphere.
+         */
+        float Z;
+        /** @brief A threshold for the duplex point elimination.
+         */
+        float diff;
+        /** @brief Temp camera position for duplex position elimination.
+         */
+        std::vector<cv::Point3d> CameraPos_temp;
+        /** @brief Make all view points having the some distance from the focal point used by the camera view.
```

Comment:
replace:
Make all view points having the same distance from the focal point used by the camera view.